kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -0,0 +1,173 @@
+    public static final int DEFAULT_WRITE_VERSION = 2;
+
+    private final UUID txnId = UUID.randomUUID();
+
+    private final boolean isNewTable; // the transaction is creating a new table
+    private final String engineInfo;
+    private final Operation operation;
+    private final Path dataPath;
+    private final Path logPath;
+    private final Protocol protocol;
+    private final Metadata metadata;
+    private final SnapshotImpl readSnapshot;
+
+    private boolean closed; // To avoid trying to commit the same transaction again.
+
+    public TransactionImpl(
+            boolean isNewTable,
+            Path dataPath,
+            Path logPath,
+            SnapshotImpl readSnapshot,
+            String engineInfo,
+            Operation operation,
+            Protocol protocol,
+            Metadata metadata) {
+        this.isNewTable = isNewTable;
+        this.dataPath = dataPath;
+        this.logPath = logPath;
+        this.readSnapshot = readSnapshot;
+        this.engineInfo = engineInfo;
+        this.operation = operation;
+        this.protocol = protocol;
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Row getTransactionState(Engine engine) {
+        return TransactionStateRow.of(metadata, dataPath.toString());
+    }
+
+    @Override
+    public List<String> getPartitionColumns(Engine engine) {
+        return VectorUtils.toJavaList(metadata.getPartitionColumns());
+    }
+
+    @Override
+    public StructType getSchema(Engine engine) {
+        return readSnapshot.getSchema(engine);
+    }
+
+    @Override
+    public TransactionCommitResult commit(
+            Engine engine,
+            CloseableIterable<Row> dataActions)
+            throws ConcurrentWriteException {
+        checkState(
+                !closed,
+                "Transaction is already attempted to commit. Create a new transaction.");
+        List<Row> metadataActions = new ArrayList<>();
+        metadataActions.add(createCommitInfoSingleAction(generateCommitAction()));
+        if (isNewTable) {
+            // In the future, we need to add metadata and action when there are any changes to them.
+            metadataActions.add(createMetadataSingleAction(metadata.toRow()));
+            metadataActions.add(createProtocolSingleAction(protocol.toRow()));
+        }
+
+        try (CloseableIterator<Row> stageDataIter = dataActions.iterator()) {
+            // Create a new CloseableIterator that will return the metadata actions followed by the
+            // data actions.
+            CloseableIterator<Row> dataAndMetadataActions =
+                    toCloseableIterator(metadataActions.iterator()).combine(stageDataIter);
+
+            try {
+                long readVersion = readSnapshot.getVersion(engine);
+                if (readVersion == -1) {
+                    // New table, create a delta log directory
+                    if (!engine.getFileSystemClient().mkdirs(logPath.toString())) {
```

Comment:
What happens if the directory already exists?

Reply:
returns true.